charms/layers/apache-nifi/reactive/apache_nifi.py

```diff
@@ -0,0 +1,98 @@
+    set_state('apache-nifi.installed')
+
+
+@when_all('zookeeper.joined', 'apache-nifi.installed')
+@when_not('zookeeper.ready')
+def zookeeper_wait(zookeeper):
+    hookenv.status_set('waiting', 'Waiting for Zookeeper to become available')
+
+
+@when_all('zookeeper.ready', 'apache-nifi.installed')
+@when_not('apache-nifi.cluster')
+def zookeeper_config(zookeeper):
+    hookenv.status_set('maintenance', 'Changing Apache NiFi to run as a cluster')
+    hookenv.log('Adding Apache Zookeeper -- Changing Apache NiFi to run as a cluster')
+    conf = hookenv.config()
+    zookeeper_servers_string = ''
+    for zk_unit in zookeeper.zookeepers():
+        zookeeper_servers_string += '%s:%s,' % (zk_unit['host'], zk_unit['port'])
+    re_edit_in_place('%s/files/nifi-1.0.0/conf/nifi.properties' % hookenv.charm_dir(), {
+        r'.*nifi.cluster.is.node.*': 'nifi.cluster.is.node=true',
+        r'.*nifi.cluster.node.address.*': 'nifi.cluster.node.address=%s' % hookenv.unit_public_ip(),
+        r'.*nifi.web.http.port.*': 'nifi.web.http.port=%s' % conf['nifi-port'],
+        r'.*nifi.cluster.node.protocol.port.*': 'nifi.cluster.node.protocol.port=%s' % conf['cluster-port'],
+        r'.*nifi.zookeeper.connect.string.*': 'nifi.zookeeper.connect.string=%s' % zookeeper_servers_string
+    })
+    service_restart('nifi')
+    set_state('apache-nifi.cluster')
+    hookenv.status_set('active', 'Apache NiFi is running in a cluster with Apache Zookeeper')
+
+
+@when_all('zookeeper.ready', 'apache-nifi.installed', 'apache-nifi.cluster')
+def zookeeper_changed(zookeeper):
+    hookenv.log('Checking if Zookeeper has changed')
+    zookeeper_servers_string = ''
+    filesdir = '%s/files' % hookenv.charm_dir()
+    for zk_unit in zookeeper.zookeepers():
+        zookeeper_servers_string += '%s:%s,' % (zk_unit['host'], zk_unit['port'])
+    if zookeeper_servers_string[:-1] not in open('%s/nifi-1.0.0/conf/nifi.properties' % filesdir).read():
+        hookenv.status_set('maintenance', 'Zookeeper has changed. Updating Apache NiFi settings and restarting')
+        re_edit_in_place('%s/nifi-1.0.0/conf/nifi.properties' % filesdir, {
+            r'.*nifi.zookeeper.connect.string.*': 'nifi.zookeeper.connect.string=%s' % zookeeper_servers_string[:-1]
+        })
+        service_restart('nifi')
+        hookenv.status_set('active', 'Apache NiFi is running in a cluster with Apache Zookeeper')
+
+
+@when('apache-nifi.cluster')
+@when_not('zookeeper.joined', 'zookeeper.ready')
+def zookeeper_removed():
+    hookenv.status_set('maintenance', 'Removing Apache NiFi from cluster')
+    re_edit_in_place('%s/files/nifi-1.0.0/conf/nifi.properties' % hookenv.charm_dir(), {
+        r'.*nifi.cluster.is.node.*': 'nifi.cluster.is.node=false'
+    })
+    service_stop('nifi')
+    remove_state('apache-nifi.cluster')
+    hookenv.status_set('waiting', 'Apache NiFi is configured for standalone mode, but not running')
```

Comment:
A user reading this message should know what to do to get it running again: `Stopped (waiting for zookeeper)`